common/sensor/dev/intel_peci.c

```diff
@@ -0,0 +1,180 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include "sensor.h"
+#include "pal.h"
+#include <drivers/peci.h>
+#include "hal_peci.h"
+#include <string.h>
+#include "intel_peci.h"
+
+#define READ_RTY 0x03
+
+#define RDPKG_IDX_PKG_TEMP 0x02
+#define RDPKG_IDX_DIMM_TEMP 0x0E
+#define RDPKG_IDX_TJMAX_TEMP 0x10
+
+static bool peci_rty_read(uint8_t cmd, uint8_t addr, uint8_t idx, uint16_t param, uint8_t rlen, uint8_t *rbuf)
+{
+  if (!rbuf)
+    return false;
+
+  int i;
+  for (i = 0; i < READ_RTY; i++) {
+    if (!peci_read(cmd, addr, idx, param, rlen, rbuf)) {
+      const uint8_t cc = rbuf[0];
+
+      if (cc == PECI_CC_RSP_SUCCESS)
+        break;
+    }
+    
+    /* TODO: read failed, maybe add delay before next read */
+  }
+
+  return (i == READ_RTY) ? false : true;
+}
+
+static bool get_cpu_tjmax(uint8_t addr, int *reading)
+{
+  if (!reading)
+    return false;
+
+  const uint16_t param = 0x00;
+  const uint8_t rlen = 0x05;
+  uint8_t rbuf[rlen];
+  memset(rbuf, 0, sizeof(rbuf));
+
+  if (peci_rty_read(PECI_CMD_RD_PKG_CFG0, addr, RDPKG_IDX_TJMAX_TEMP, param, rlen, rbuf) == false)
+    return false;
+
+  sen_val *sval = (sen_val *)reading;
+  sval->integer = rbuf[3];
+  return true;
+}
+
+static bool get_cpu_margin(uint8_t addr, int *reading)
+{
+  if (!reading)
+    return false;
+
+  const uint16_t param = 0xFF;
+  const uint8_t rlen = 0x05;
+  uint8_t rbuf[rlen];
+  memset(rbuf, 0, sizeof(rbuf));
+
+  if (peci_rty_read(PECI_CMD_RD_PKG_CFG0, addr, RDPKG_IDX_PKG_TEMP, param, rlen, rbuf) == false)
+    return false;
+
+  sen_val *sval = (sen_val *)reading;
+  sval->integer = ((int16_t)((rbuf[2] << 8) | rbuf[1]) >> 6) + 1;
+  return true;
+}
+
+static bool get_cpu_pwr(uint8_t sen_num, int *reading)
+{
+  if (!reading)
+    return false;
+
+  int pwr = 0;
+  if (peci_getPwr(sen_num, &pwr) == false)
+    return false;
+
+  sen_val *sval = (sen_val *)reading;
+  sval->integer = (int16_t)pwr;
+  return true;
+}
+
+static bool get_cpu_temp(uint8_t addr, int *reading)
+{
+  if (!reading)
+    return false;
+
+  sen_val tjmax = {0};
+  if (get_cpu_tjmax(addr, (int *)&tjmax) == false)
+    return false;
+
+  sen_val margin = {0};
+  if (get_cpu_margin(addr, (int *)&margin) == false)
+    return false;
+
+  sen_val *sval = (sen_val *)reading;
+  sval->integer = tjmax.integer + margin.integer;
+  return true;
+}
+
+static bool get_dimm_temp(uint8_t addr, uint8_t type, int *reading)
+{
+  if (!reading)
+    return false;
+
+  const uint16_t param = (type == PECI_TEMP_DIMM_A) ? 0x00 :
```

Comment:
This makes assumptions on DIMM_X always being mapped to DIMM_A through DIMM_H on platforms.  With the recent proposal by the RTP team to renumber DIMM slots to A0 through A7, hard-coding these values might lead to future code changes to accommodate both numberings.

Reply:
Thanks for the suggestion, I think we can follow the intel PECI command format to define which value should be used in the PECI generic code.
In the PECI specification, the DDR DIMM temperature command with
&nbsp; request parameters - channel index, and
&nbsp; response data - DIMM# temperature.
There are 8 channels in this CPU SKU with two DIMMs per channel.
Therefore, the DIMM temperature of the enum can be defined as below.

      PECI_TEMP_CHANNEL0_DIMM0,
      PECI_TEMP_CHANNEL0_DIMM1,
      PECI_TEMP_CHANNEL1_DIMM0,
      PECI_TEMP_CHANNEL1_DIMM1,
      PECI_TEMP_CHANNEL2_DIMM0,
      PECI_TEMP_CHANNEL2_DIMM1,
      PECI_TEMP_CHANNEL3_DIMM0,
      PECI_TEMP_CHANNEL3_DIMM1,
      PECI_TEMP_CHANNEL4_DIMM0,
      PECI_TEMP_CHANNEL4_DIMM1,
      PECI_TEMP_CHANNEL5_DIMM0,
      PECI_TEMP_CHANNEL5_DIMM1,
      PECI_TEMP_CHANNEL6_DIMM0,
      PECI_TEMP_CHANNEL6_DIMM1,
      PECI_TEMP_CHANNEL7_DIMM0,
      PECI_TEMP_CHANNEL7_DIMM1,



In the platform code, it can map the DIMM temperature sensor to the physical DIMM channel with any naming by the above enum define.
I think it is more clear in the common code and separate the platform naming rule.



I am looking forward to your reply.
Thanks